src/replication.c

```diff
@@ -750,6 +766,225 @@ is_replica_newly_connected(spec_t *spec, replica_t *new_replica)
+
+	replica_count = spec->healthy_rcount + spec->degraded_rcount;
+
+	if (replica_count >= spec->replication_factor) {
+		ISTGT_ERRLOG("known replica count %d is greater "
+		    "than or equal to replciation factor %d\n",
+		    replica_count, spec->replication_factor);
+		return -1;
+	}
+
+	jobj = json_object_new_object();
+	json_object_object_add(jobj, "replicaId",
+	    json_object_new_string(rep->replica_id));
+	json_object_object_add(jobj, "replicaZvolGuid",
+	    json_object_new_uint64(rep->zvol_guid));
+	json_object_object_add(jobj, "volumeName",
+	    json_object_new_string(spec->volname));
+	json_string = json_object_to_json_string_ext(jobj, JSON_C_TO_STRING_PLAIN);
+
+	data_len = strlen(json_string) + 1;
+	data = (char *) malloc(sizeof(char *)*data_len);
+	memset(data, 0, data_len);
+	strncpy(data, json_string, data_len);
+
+	rc = update_volume_config(data);
+	if (rc < 0) {
+		ISTGT_ERRLOG("failed to update replica(%s:%lu) to known replica list\n",
+		    rep->replica_id, rep->zvol_guid);
+	} else {
+		(void) update_in_memory_known_replica_list(spec, rep);
+	}
+
+	free(data);
+	json_object_put(jobj);
+	return rc;
+}
+
+/* CHECK_IS_IT_KNOWN_REPLICA returns true if replica_id and replica_zvol_guid
+ * is matched with any of known replicas replica_id and replica_zvol_guid
+ */
+#define CHECK_IS_IT_KNOWN_REPLICA(_spec, _rep, val)				\
+	do {									\
+		val = false;							\
+		trusty_replica_t *knr;						\
+		TAILQ_FOREACH(knr, &_spec->lu->known_replica, next){	\
+			if (strcmp(knr->replica_id, _rep->replica_id) == 0 &&	\
+			    knr->zvol_guid == _rep->zvol_guid) {		\
+				val = true;					\
+				break;						\
+			}							\
+		}								\
+	}while(0)
+
+/*
+ * is_known_replica returns true whether the replica is known or not based
+ * on below factors
+ * return value true --> known replica
+ * return value false --> unknown replica
+ * 1) If quorum is off, then treat as unknown replica.
+ * 2) If known replica count is less than replication factor
+ *    return true(newly created volume/upgrade scenarios).
+ * 3) Check whether replicaId and zvol guid of replica
+ *    is present in known replica list return based on
+ *    that condition
+ */
+static bool
+is_known_replica(spec_t *spec, replica_t *new_replica)
+{
+	int known_replica_count;
+	bool result = true;
+
+	// Replica scale up case and replica movement cases
+	if (new_replica->quorum == 0) {
+		return false;
+	}
+	// This is to handle upgrade scenarios and newely
+	// connected volume
+	known_replica_count = get_known_replica_count(spec);
+	if (known_replica_count < spec->replication_factor) {
+		return result;
+	}
+	// Check If replica is present in spec->lu->know_replica_list
+	// then return true else false
+	CHECK_IS_IT_KNOWN_REPLICA(spec, new_replica, result);
+	return result;
+}
+
+/* is_replica_exist returns true if replica_id is present
+ * in known replica list else return false.
+ */
+static int
+is_replica_exist(spec_t *spec, char *replica_id) {
```

Comment:
make this function as `is_trusted_replicae_exist"

Reply:
or `is_trusted_replicas_contain_replicaid`

Reply:
How about `is_trusted_replica_id_exist`?